include/util/geojson_debug_logger.hpp

```diff
@@ -0,0 +1,154 @@
+    // to be later filled with additional data
+    static bool Open(const std::string &logfile)
+    {
+        // if a file is open, close it off. When this function returns, there should be an open
+        // logfile. However, there is a brief period between closing and openen where we could miss
+        // out on log output. Such a sad life
+        if (ofs.is_open())
+            Close();
+
+        // make sure to syncronize logging output, cannot be locked earlier, since Close also locks
+        // and we don't want deadlocks
+        std::lock_guard<std::mutex> guard(lock);
+        ofs.open(logfile, std::ios::binary);
+
+        // set up a feature collection
+        ofs << "{\n\t\"type\": \"FeatureCollection\",\n\t\"features\": [\n\t";
+        // remember whether we need to output a colon
+        first = true;
+
+        return static_cast<bool>(ofs);
+    }
+
+    // finalising touches on the GeoJson
+    static bool Close()
+    {
+        // make sure to syncronize logging output
+        std::lock_guard<std::mutex> guard(lock);
+
+        // finishe the geojson feature collection and close it all off
+        if (ofs.is_open())
+        {
+            ofs << "\n\t]\n}";
+            ofs.close();
+        }
+
+        return static_cast<bool>(ofs);
+    }
+
+    static void SetPolicy(geojson_conversion_policy *new_policy) { policy = new_policy; }
+
+  private:
+    static bool first;
+    static std::mutex lock;
+    static std::ofstream ofs;
+
+    static geojson_conversion_policy *policy;
+};
+
+// make sure to do opening and closing of our guard
+template <class geojson_conversion_policy, LoggingScenario scenario = LoggingScenario::eDefault>
+class ScopedGeojsonLoggerGuard
+{
+  public:
+    template <typename... Args>
+    ScopedGeojsonLoggerGuard(const std::string logfile, Args &&... args)
+        : policy(std::forward<Args>(args)...)
+    {
+        GeojsonLogger<geojson_conversion_policy, scenario>::Open(logfile);
+        GeojsonLogger<geojson_conversion_policy, scenario>::SetPolicy(&policy);
+    }
+
+    ~ScopedGeojsonLoggerGuard()
+    {
+        GeojsonLogger<geojson_conversion_policy, scenario>::Close();
+        GeojsonLogger<geojson_conversion_policy, scenario>::SetPolicy(nullptr);
+    }
+
+  private:
+    geojson_conversion_policy policy;
+};
+
+// make sure to define our local template variables
+template <class geojson_conversion_policy, LoggingScenario scenario>
+bool GeojsonLogger<geojson_conversion_policy, scenario>::first;
+
+template <class geojson_conversion_policy, LoggingScenario scenario>
+std::mutex GeojsonLogger<geojson_conversion_policy, scenario>::lock;
+
+template <class geojson_conversion_policy, LoggingScenario scenario>
+std::ofstream GeojsonLogger<geojson_conversion_policy, scenario>::ofs;
+
+template <class geojson_conversion_policy, LoggingScenario scenario>
+geojson_conversion_policy* GeojsonLogger<geojson_conversion_policy, scenario>::policy;
```

Comment:
Hmm don't we have to define them in a single impl. file - otherwise we would validate the one-definition-rule when multiple files include this header?